### PR TITLE
vcs detection

### PR DESCRIPTION
**Pull request by Raruto**
In response to: https://github.com/afragen/local-development/issues/4, [`WP_Automatic_Updater::is_vcs_checkout`](https://developer.wordpress.org/reference/classes/wp_automatic_updater/is_vcs_checkout/) is a tree walker function.

_.. it breaks all if a parent folder is under vcs .._

---

In particular, in this version of mine:
- automatically added checkboxes are disabled by default
- a visual feedback is added to the admin bar when working on localhost
- increased visibility of the "in local development" label

### CODE REVIEW

**Comment by afragen** [comment]
Can we split these into separate PRs. I have to see what the styling changes look like. Also, it's better to add the CSS using `wp_enqueue_style()`. Put the CSS in `assets/css`.

File: src/Local_Development/Base.php [hunk 1]
```diff
@@ -93,7 +93,7 @@ public function row_meta( $links, $file ) {
 		if ( ( ! empty( self::$plugins ) && array_key_exists( $file, self::$plugins ) ) ||
 			( ! empty( self::$themes ) && array_key_exists( $file, self::$themes ) )
 		) {
-			$links[] = '<strong>' . self::$message . '</strong>';
+			$links[] = '<strong style="color:red; font-weight:700;">' . self::$message . '</strong>';
```
Comment (afragen) [inline src/Local_Development/Base.php:96]: Is there some benefit to adding the font-weight when it’s already using the strong tag?

Any specific reason for the coloring?

File: src/Local_Development/Init.php [hunk 2]
```diff
@@ -51,22 +51,25 @@ function () use ( $config ) {
...
 			foreach ( array_keys( $plugins_themes[ $type ] ) as $file ) {
 				$wp_path  = 'plugins' === $type ? wp_normalize_path( WP_PLUGIN_DIR ) : wp_normalize_path( get_theme_root() );
 				$slug     = 'plugins' === $type ? dirname( $file ) : $file;
 				$filepath = "$wp_path/$slug";
-				$is_vcs   = $updater->is_vcs_checkout( $filepath );
-				if ( $is_vcs ) {
-					$config[ $type ][ $file ] = '1';
+
+				foreach ( $vcs_dirs as $vcs_dir ) {
+					$is_vcs = @is_dir( rtrim( $filepath, '\\/' ) . "/$vcs_dir" );
```
Comment (afragen) [inline src/Local_Development/Init.php]: Is the rtrim any different than untrailingslashit()?

I think it would make the code more readable.

**Comment by Raruto** [comment]
> Can we split these into separate PRs.

Ok, done.

> Also, it's better to add the CSS using wp_enqueue_style(). Put the CSS in assets/css.

This is generally true but it's not error-proof.

Sometimes, **especially during development**, it can happen that assets files are not correctly loaded by browser (with the inline strategy this simply cannot happen).

> I have to see what the styling changes look like.

The aesthetic modification is really small (**3 css rules**), but I consider it very effective (to easily distinguish local instances). Anyway, I had also added a checkbox in extras to disable it in case you don't want it ...

```css
#wpadminbar #wp-admin-bar-site-name > .ab-item::after {
  content: " - localhost";
  font-weight: 800;
  font-family: Monospace;
  color: #fff;
}

#wpadminbar #wp-admin-bar-site-name > .ab-item {
  background-color: #008000;
  color: #ff0;
}

#wpadminbar #wp-admin-bar-site-name > .ab-item::before {
  background: url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIGFyaWEtaGlkZGVuPSJ0cnVlIiB3aWR0aD0iMTQiIGhlaWdodD0iMTYiIHN0eWxlPSItbXMtdHJhbnNmb3JtOnJvdGF0ZSgzNjBkZWcpOy13ZWJraXQtdHJhbnNmb3JtOnJvdGF0ZSgzNjBkZWcpO3RyYW5zZm9ybTpyb3RhdGUoMzYwZGVnKSI+PHBhdGggZmlsbC1ydWxlPSJldmVub2RkIiBkPSJNOS41IDNMOCA0LjUgMTEuNSA4IDggMTEuNSA5LjUgMTMgMTQgOCA5LjUgM3ptLTUgMEwwIDhsNC41IDVMNiAxMS41IDIuNSA4IDYgNC41IDQuNSAzeiIgZmlsbD0iI2ZmMCIvPjwvc3ZnPg==) center center no-repeat !important;
  content: " " !important;
  display: block;
  width: 16px;
  height: 21px;
}
```

File: src/Local_Development/Base.php (hunk 1 above)
Comment (Raruto) [reply src/Local_Development/Base.php:96]: I would have preferred to use standard typographic wordpress classes, but I don't think they currently exist...

> Is there some benefit to adding the font-weight when it’s already using the strong tag?

You did well, this change only makes it a little bit more visible (bolder).

> Any specific reason for the coloring?

- currently, when there are many plugins installed it is not always easy to see which plugins are under development
- this plugin must help the developer to easily recognize which plugins are currently under development (for this you have to use an accent color, which by definition differs from the default one)
- core wordpress features are disabled (a different color better explains unexpected behaviors)

File: src/Local_Development/Init.php (hunk 2 above)
Comment (Raruto) [reply src/Local_Development/Init.php]: That part is based on the current implementation of [is_vcs_checkout](https://developer.wordpress.org/reference/classes/wp_automatic_updater/is_vcs_checkout/)

Looking at [untrailingslashit](https://developer.wordpress.org/reference/functions/untrailingslashit/) implementation they seem to be equivalent (I'll change it)

Comment (Raruto) [reply src/Local_Development/Init.php]: Wait, there is a little difference:
- `rtrim( $string, '\\/' )` <-- is_vcs_checkout
- `rtrim( $string, '/\\' )` <-- untrailingslashit

But exactly what changes?

Comment (Raruto) [reply src/Local_Development/Init.php]: I think they should be truly equivalent (edit done).

Although it is somewhat strange that within [is_vcs_checkout](https://developer.wordpress.org/reference/classes/wp_automatic_updater/is_vcs_checkout/) are used both...

Comment (afragen) [reply src/Local_Development/Init.php]: WP core is spaghetti.

**Comment by afragen** [comment]
Can you please make separate PRs for the CSS rules and for the icons. Both should be selectable options. If you can find/create a Gitea icon that would be great. I think differentiating public and private repos isn't necessary.

**Comment by afragen** [comment]
I just realized I have inline css in here too. Go ahead and make the PR that way.

**Pull request merged**